arangod/Aql/LateMaterializedOptimizerRulesCommon.cpp

```diff
@@ -0,0 +1,120 @@
+////////////////////////////////////////////////////////////////////////////////
+/// DISCLAIMER
+///
+/// Copyright 2019 ArangoDB GmbH, Cologne, Germany
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Yuriy Popov
+////////////////////////////////////////////////////////////////////////////////
+
+#include "Aql/Ast.h"
+#include "Aql/LateMaterializedOptimizerRulesCommon.h"
+
+using namespace arangodb::aql;
+
+namespace {
+
+// traverse the AST, using previsitor
+void traverseReadOnly(AstNode* node, AstNode* parentNode, size_t childNumber,
```

Comment:
you can just use `Ast::traverseReadOnly(AstNode* node, std::function<...> const& preVisitor, std::function<...> const& postVisitor);`

Reply:
I need parentNode and childNumber. Ast::traverseReadOnly does not provide this information.